subsys/net/lib/ftp_client/src/ftp_client.c

```diff
@@ -0,0 +1,793 @@
+	int ret = FTP_CODE_500;
+
+	sprintf(code_str, "%d ", success_code);
+	if (strstr(message, code_str)) {
+		ret = success_code;
+	}
+
+	return ret;
+}
+
+int establish_data_channel(const char *pasv_msg)
+{
+	int ret;
+	char tmp[16];
+	char *tmp1, *tmp2;
+	u16_t data_port;
+	int data_sock;
+
+	/* Parse Server port from passive message
+	 * e.g. "227 Entering Passive Mode (90,130,70,73,86,111)"
+	 */
+	tmp1 = strstr(pasv_msg, "(");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	tmp2++;
+	tmp1 = strstr(tmp2, ",");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	tmp2++;
+	tmp1 = strstr(tmp2, ",");
+	tmp1++;
+	tmp2 = strstr(tmp1, ",");
+	memset(tmp, 0x00, 16);
+	strncpy(tmp, (const char *)tmp1, (size_t)(tmp2 - tmp1));
+	data_port = atoi(tmp) << 8;
+	tmp2++;
+	tmp1 = strstr(tmp2, ")");
+	memset(tmp, 0x00, 16);
+	strncpy(tmp, (const char *)tmp2, (size_t)(tmp1 - tmp2));
+	data_port += atoi(tmp);
+
+	/* Establish the second connect for data */
+	if (sec_tag <= 0) {
+		data_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	} else {
+		data_sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	}
+	if (data_sock < 0) {
+		LOG_ERR("socket(data) failed: %d", -errno);
+		ret = -errno;
+	}
+
+	if (sec_tag > 0) {
+		sec_tag_t sec_tag_list[1] = { sec_tag };
+
+		ret = setsockopt(data_sock, SOL_TLS, TLS_SEC_TAG_LIST,
+				sec_tag_list, sizeof(sec_tag_t));
+		if (ret) {
+			LOG_ERR("set tag list failed: %d", -errno);
+			close(data_sock);
+			return -errno;
+		}
+
+	}
+	m_remote.sin_port = htons(data_port);
+	ret = connect(data_sock, (struct sockaddr *)&m_remote,
+		 sizeof(struct sockaddr_in));
+	if (ret < 0) {
+		LOG_ERR("connect(data) failed: %d", -errno);
+		close(data_sock);
+		return -errno;
+	}
+
+	return data_sock;
+}
+
+/**@brief Send FTP message via socket
+ */
+int do_ftp_send_ctrl(const u8_t *message, int length)
+{
+	int ret;
+	u32_t offset = 0;
+
+	LOG_HEXDUMP_DBG(message, length, "TXC");
+
+	while (offset < length) {
+		ret = send(client.sock, message + offset, length - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send(ctrl) failed: %d", -errno);
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+		ret = 0;
+	}
+
+	LOG_DBG("CMD sent");
+	return ret;
+}
+
+/**@brief Send FTP data via socket
+ */
+int do_ftp_send_data(const char *pasv_msg, u8_t *message, u16_t length)
+{
+	int data_sock;
+	int ret;
+	u32_t offset = 0;
+
+	/* Establish data channel */
+	ret = establish_data_channel(pasv_msg);
+	if (ret <= 0) {
+		return ret;
+	}
+
+	LOG_HEXDUMP_DBG(message, length, "TXD");
+
+	data_sock = ret;
+	while (offset < length) {
+		ret = send(data_sock, message + offset, length - offset, 0);
+		if (ret < 0) {
+			LOG_ERR("send(data) failed: %d", -errno);
+			ret = -errno;
+			break;
+		}
+		offset += ret;
+		ret = 0;
+	}
+
+	LOG_DBG("DATA sent");
+	close(data_sock);
+	return ret;
+}
+
+/**@brief Receive FTP message from socket
+ */
+int do_ftp_recv_ctrl(bool post_result, int success_code)
+{
+	int ret;
+	struct pollfd fds[1];
+
+	memset(ctrl_buf, 0x00, NET_IPV4_MTU);
```

Comment:
not needed any more, removed.

Reply:
After calling do_ftp_recv_ctrl(), content of ctrl_buf is checked. In case the caller not check the return code (e.g. recv failed scenario), make sure the buffer is empty.